### PR TITLE
Bug 1712920 - Implement rate limiting in ping upload

### PR DESCRIPTION
**Pull request by brizental**
This actually fixes a bug: there was no code path to re-start the PingUploader dispatcher once it was stopped due to max recoverable error being hit. None of our consumers have updated yet,s o if I make a release after this is merged we should be fine.

### Pull Request checklist ###

- [x] **Quality**: Make sure this PR builds and runs cleanly.
  - Inside the `glean/` folder, run:
    - `npm run test` Runs _all_ tests
    - `npm run lint` Runs _all_ linters
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry to `CHANGELOG.md` or an explanation of why it does not need one
- [x] **Documentation**: This PR includes documentation changes, an explanation of why it does not need that or a follow-up bug has been filed to do that work
  - https://github.com/mozilla/glean/pull/1758

### CODE REVIEW

File: glean/src/core/upload/index.ts [hunk 1]
```diff
@@ -121,6 +128,35 @@ class PingUploader implements PingsDatabaseObserver {
...
+      this.dispatcher.resume();
+    } else {
+      this.dispatcher.stop();
+
+      if (rateLimiterState === RateLimiterState.Throttled) {
+        log(
+          LOG_TAG,
+          [
+            "Attempted to upload a ping, but Glean is currently throttled.",
+            `Pending pings will be processed in ${(remainingTime || 0) / 1000}s.`
```
Comment (brizental) [inline glean/src/core/upload/index.ts:142]: I kinda of don't like this message, as it is promising something that is not true: the pings _may_ be processed after the remaining time, but it doesn't mean they will, there is no trigger that goes off once that time is over. 

On the other SDKs we do have one though, so I will file a bug to do the same here. See how that happens in the [Python](https://github.com/mozilla/glean/blob/main/glean-core/python/glean/net/ping_upload_worker.py#L156) or [Android](https://github.com/mozilla/glean/blob/main/glean-core/android/src/main/java/mozilla/telemetry/glean/scheduler/PingUploadWorker.kt#L120) SDKs.

Question is: is everyone confortable with the empty promise here until that is merged? I personally don't think it is a big deal, but best to flag it.

Comment (brizental) [reply glean/src/core/upload/index.ts:142]: https://bugzilla.mozilla.org/show_bug.cgi?id=1727076 filed.

Comment (Dexterp37) [reply glean/src/core/upload/index.ts:142]: > Question is: is everyone confortable with the empty promise here until that is done? I personally don't think it is a big deal, but best to flag it.

I don't think it's a big deal.

File: glean/src/core/upload/rate_limiter.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+ * An enum to represent the current state of the RateLimiter.
+ */
+export const enum RateLimiterState {
+  // The RateLimiter has not reached the maximum count and is still incrementing.
+  Incrementing,
+  // The RateLimiter has not reached the maximum count, but it is also not incrementing.
+  Stopped,
+  // The RateLimiter has reached the maximum count for the current interval.
+  //
+  // This variant contains the remaining time (in milliseconds)
```
Comment (Dexterp37) [inline glean/src/core/upload/rate_limiter.ts:17]: This is a bit confusing: the enum itself is not holding the remaining time in MS, it's just that we return the time from `getState`: could you clarify this a bit?

File: glean/src/core/upload/rate_limiter.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+import { isUndefined, getMonotonicNow } from "../utils.js";
+
+/**
+ * An enum to represent the current state of the RateLimiter.
+ */
+export const enum RateLimiterState {
+  // The RateLimiter has not reached the maximum count and is still incrementing.
+  Incrementing,
+  // The RateLimiter has not reached the maximum count, but it is also not incrementing.
+  Stopped,
```
Comment (Dexterp37) [inline glean/src/core/upload/rate_limiter.ts:14]: We seem to be returning the remaining time in this case, too: should we amend this comment or change `getState`?

File: glean/src/core/upload/rate_limiter.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+    } {
+    if (this.shouldReset()) {
+      this.reset();
+    }
+
+    const remainingTime = this.interval - this.elapsed;
+    if (this.stopped) {
+      return {
+        state: RateLimiterState.Stopped,
+        remainingTime,
```
Comment (Dexterp37) [inline glean/src/core/upload/rate_limiter.ts:101]: We're returning the remaining time here, do we need to fix the comment on the enum definition or remove this?

File: glean/src/core/upload/rate_limiter.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+
+  get elapsed(): number {
+    if (isUndefined(this.started)) {
+      return NaN;
+    }
+
+    const now = getMonotonicNow();
+    const elapsed = now - this.started;
+
+    // It's very unlekily elapsed will be a negative number since we are using a monotonic timer
```
Comment (Dexterp37) [inline glean/src/core/upload/rate_limiter.ts:47]:
```suggestion
    // It's very unlikely elapsed will be a negative number since we are using a monotonic timer
```

File: glean/src/core/upload/rate_limiter.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+  private reset(): void {
+    this.started = getMonotonicNow();
+    this.count = 0;
+    this.stopped = false;
+  }
+
+  /**
+   * The rate limiter should reset if
+   *
+   * 1. It has never started;
```
Comment (Dexterp37) [inline glean/src/core/upload/rate_limiter.ts:65]: Why is this the case? If it has never started, it should already be in the stopped state, right?

**Review by Dexterp37 — CHANGES_REQUESTED** [review]

File: glean/src/core/upload/rate_limiter.ts (hunk 2 above)
Comment (brizental) [reply glean/src/core/upload/rate_limiter.ts:17]: Ack, this is because I copy pasted this comment from Rust, where enums may hold data. Which is not the case for TypeScript. I'll reword it.

File: glean/src/core/upload/rate_limiter.ts (hunk 3 above)
Comment (brizental) [reply glean/src/core/upload/rate_limiter.ts:14]: Nah, I don't think we need to mention the remaining time at all in these enum comments.

File: glean/src/core/upload/rate_limiter.ts (hunk 4 above)
Comment (brizental) [reply glean/src/core/upload/rate_limiter.ts:101]: Answered by the previous responses I think.

File: glean/src/core/upload/rate_limiter.ts (hunk 6 above)
Comment (brizental) [reply glean/src/core/upload/rate_limiter.ts:65]: If it has never started it does not have a `startTime` so we need to generate it. I can reword it to say exactly this though. Full disclaimer, this was also copied from the Rust code, hehe. 

Maybe I should re-read these comments after sending the PR?!

**Review by Dexterp37 — APPROVED** [review]
